src/hotspot/share/runtime/handshake.hpp

```diff
@@ -99,8 +100,20 @@ class HandshakeState {
   bool process_self_inner();
 
   bool have_non_self_executable_operation();
-  HandshakeOperation* pop_for_self();
-  HandshakeOperation* pop();
+  HandshakeOperation* get_op_for_self();
+  HandshakeOperation* get_op();
+  void remove_op(HandshakeOperation* op);
+
+  void set_active_handshaker(Thread* thread) { Atomic::store(&_active_handshaker, thread); }
```

Comment:
Should:
```Thread* _active_handshaker;```
on L89 above be:
```Thread* volatile _active_handshaker;```

Reply:
Yes, you are correct.

The problematic scenario is Deoptimization::deoptimize_frame which is uses this method to figure out how to deopt the frame. In that case _active_handshaker can be concurrently accessed.